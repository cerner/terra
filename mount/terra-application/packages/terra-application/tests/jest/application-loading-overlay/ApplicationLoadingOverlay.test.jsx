import React from 'react';
import uuidv4 from 'uuid/v4';

import ApplicationLoadingOverlay from '../../../src/application-loading-overlay/ApplicationLoadingOverlay';
import ApplicationLoadingOverlayContext from '../../../src/application-loading-overlay/ApplicationLoadingOverlayContext';

jest.mock('uuid/v4');

describe('ApplicationLoadingOverlay', () => {
  let reactUseContext;
  let loadingOverlayContextValue;

  beforeAll(() => {
    uuidv4.mockReturnValue('test-id');

    /**
     * Until Enzyme is updated with full support for hooks, we need to
     * mock out the useContext implementation.
     */
    reactUseContext = React.useContext;
    React.useContext = (contextValue) => {
      if (ApplicationLoadingOverlayContext === contextValue) {
        return loadingOverlayContextValue;
      }
      return reactUseContext(contextValue);
    };
  });

  afterAll(() => {
    React.useContext = reactUseContext;
  });

  it('should render loading overlay as closed', () => {
    loadingOverlayContextValue = {
      show: jest.fn(),
      hide: jest.fn(),
    };

    const wrapper = mount(
      <ApplicationLoadingOverlay />,
    );

    expect(wrapper).toMatchSnapshot();

    expect(loadingOverlayContextValue.show.mock.calls.length).toBe(0);
    expect(loadingOverlayContextValue.hide.mock.calls.length).toBe(1);

    wrapper.unmount();
    expect(loadingOverlayContextValue.show.mock.calls.length).toBe(0);
    expect(loadingOverlayContextValue.hide.mock.calls.length).toBe(2);
  });

  it('should render loading overlay as open', () => {
    loadingOverlayContextValue = {
      show: jest.fn(),
      hide: jest.fn(),
    };

    const wrapper = mount(
      <ApplicationLoadingOverlay isOpen />,
    );

    expect(wrapper).toMatchSnapshot();

    expect(loadingOverlayContextValue.show.mock.calls.length).toBe(1);
    expect(loadingOverlayContextValue.show.mock.calls[0][0]).toBe('test-id');
    expect(loadingOverlayContextValue.hide.mock.calls.length).toBe(0);

    wrapper.unmount();
    expect(loadingOverlayContextValue.show.mock.calls.length).toBe(1);
    expect(loadingOverlayContextValue.hide.mock.calls.length).toBe(1);
    expect(loadingOverlayContextValue.hide.mock.calls[0][0]).toBe('test-id');
  });

  it('should transition from open to closed', () => {
    loadingOverlayContextValue = {
      show: jest.fn(),
      hide: jest.fn(),
    };

    const wrapper = mount(
      <ApplicationLoadingOverlay isOpen />,
    );

    expect(wrapper).toMatchSnapshot();

    expect(loadingOverlayContextValue.show.mock.calls.length).toBe(1);
    expect(loadingOverlayContextValue.show.mock.calls[0][0]).toBe('test-id');
    expect(loadingOverlayContextValue.hide.mock.calls.length).toBe(0);

    wrapper.setProps({ isOpen: false });

    expect(loadingOverlayContextValue.show.mock.calls.length).toBe(1);
    expect(loadingOverlayContextValue.hide.mock.calls.length).toBe(2);
    expect(loadingOverlayContextValue.hide.mock.calls[0][0]).toBe('test-id');
  });

  it('should redisplay loading overlay with new props', () => {
    loadingOverlayContextValue = {
      show: jest.fn(),
      hide: jest.fn(),
    };

    const wrapper = mount(
      <ApplicationLoadingOverlay isOpen />,
    );

    expect(wrapper).toMatchSnapshot();

    expect(loadingOverlayContextValue.show.mock.calls.length).toBe(1);
    expect(loadingOverlayContextValue.show.mock.calls[0][0]).toBe('test-id');
    expect(loadingOverlayContextValue.hide.mock.calls.length).toBe(0);

    wrapper.setProps({ backgroundStyle: 'dark' });

    expect(loadingOverlayContextValue.show.mock.calls.length).toBe(2);
    expect(loadingOverlayContextValue.show.mock.calls[1][0]).toBe('test-id');
    expect(loadingOverlayContextValue.hide.mock.calls.length).toBe(1);
    expect(loadingOverlayContextValue.hide.mock.calls[0][0]).toBe('test-id');
  });

  it('should honor backgroundStyle prop', () => {
    loadingOverlayContextValue = {
      show: jest.fn(),
      hide: jest.fn(),
    };

    const wrapper = mount(
      <ApplicationLoadingOverlay isOpen backgroundStyle="clear" />,
    );

    expect(wrapper).toMatchSnapshot();

    wrapper.setProps({ backgroundStyle: 'light' });
    expect(wrapper).toMatchSnapshot();

    wrapper.setProps({ backgroundStyle: 'dark' });
    expect(wrapper).toMatchSnapshot();
  });
});
