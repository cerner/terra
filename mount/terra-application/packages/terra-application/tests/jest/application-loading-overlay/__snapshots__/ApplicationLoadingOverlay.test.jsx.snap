// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`ApplicationLoadingOverlay should honor backgroundStyle prop 1`] = `
<ApplicationLoadingOverlay
  backgroundStyle="clear"
  isOpen={true}
/>
`;

exports[`ApplicationLoadingOverlay should honor backgroundStyle prop 2`] = `
<ApplicationLoadingOverlay
  backgroundStyle="light"
  isOpen={true}
/>
`;

exports[`ApplicationLoadingOverlay should honor backgroundStyle prop 3`] = `
<ApplicationLoadingOverlay
  backgroundStyle="dark"
  isOpen={true}
/>
`;

exports[`ApplicationLoadingOverlay should redisplay loading overlay with new props 1`] = `
<ApplicationLoadingOverlay
  backgroundStyle="clear"
  isOpen={true}
/>
`;

exports[`ApplicationLoadingOverlay should render loading overlay as closed 1`] = `
<ApplicationLoadingOverlay
  backgroundStyle="clear"
/>
`;

exports[`ApplicationLoadingOverlay should render loading overlay as open 1`] = `
<ApplicationLoadingOverlay
  backgroundStyle="clear"
  isOpen={true}
/>
`;

exports[`ApplicationLoadingOverlay should transition from open to closed 1`] = `
<ApplicationLoadingOverlay
  backgroundStyle="clear"
  isOpen={true}
/>
`;
